src/app/src/main/kotlin/org/imperial/mrc/hint/Extensions.kt

```diff
@@ -16,9 +25,39 @@ fun Response.asResponseEntity(): ResponseEntity<String> {
         404 -> HttpStatus.NOT_FOUND
         else -> HttpStatus.INTERNAL_SERVER_ERROR
     }
+}
+
+fun headersToMultiMap(headers: Headers): MultiValueMap<String, String> {
+    val result = LinkedMultiValueMap<String, String>()
+    headers.entries.forEach {
+        result.addAll(it.key, it.value.toList());
+    }
+    return result;
+}
+
+fun Response.asResponseEntity(): ResponseEntity<String> {
+    val httpStatus = httpStatusFromCode(this.statusCode)
 
     val body = this.body().asString("application/json")
     return ResponseEntity.status(httpStatus)
             .contentType(MediaType.APPLICATION_JSON)
             .body(body)
 }
+
+fun DownloadRequest.asStreamingResponseEntity(): ResponseEntity<StreamingResponseBody> {
+
+    val responseBody = StreamingResponseBody { outputStream: OutputStream ->
```

Comment:
Having read the Fuel docs, this seems more what needs to be done. 
https://github.com/kittinunf/fuel/tree/master/fuel#download-response-to-output-file-or-outputstream

What's not ideal is that I'm calling .response() on the DownloadRequest twice, once in order to get the headers and httpStatus to put into the ResponseEntity, once inside the StreamingResponseBody lambda in order to specify the output stream as the destination for the download. So I guess this is going to make two separate requests. Am I missing something obvious? Is there some clever async way to avoid this..?

Reply:
Making an extra HEAD request would be inexpensive seeing as it's just between hint and hintr which are on the same network, but this does seem like a pretty unintuitive interface. Had a little fiddle and this seems to do the trick and be somewhat easier to reason about:
```
        val url = URL("$baseUrl/download/spectrum/${id}")
        val conn = url.openConnection() as HttpURLConnection

        val responseEntity = ResponseEntity.status(conn.responseCode)
        responseEntity.headers(HttpHeaders(conn.headerFields as MultiValueMap<String, String>))
        responseEntity.body(StreamingResponseBody { outputStream: OutputStream ->
            conn.inputStream.use {
                it.copyTo(outputStream)
            }
        })

        return responseEntity.build()
```

Reply:
Having said this the test seems to be taking way longer to run with this code, I can't see why? :thinking:

Reply:
Ok, scratch this, this doesn't seem to be working at all.

Btw another useful check to add in the integration test is
```assertThat(response.body!!.size.toString()).isEqualTo(contentLength)```

Reply:
Ok, so the following works, not wrapping the `treamingResponseBody` in a response entity but just writing directly to the `HttpServletResponse`. But now I've come back around and thinking maybe the `Fuel` syntax isn't that bad, it's very terse after all, and making an extra request for headers should be inexpensive :woman_shrugging: 
```
    @GetMapping("/spectrum/{id}")
    @ResponseBody
    fun getSpectrum(@PathVariable("id")id: String, response: HttpServletResponse): StreamingResponseBody {

        return StreamingResponseBody { outputStream: OutputStream ->
            val url = URL("http://localhost:8888/download/spectrum/${id}")
            val conn = url.openConnection() as HttpURLConnection
            response.setContentLength(conn.contentLength)
            response.contentType = conn.contentType
            response.status = conn.responseCode
            conn.inputStream.use {
                it.copyTo(outputStream)
            }
        }
    }
```

Reply:
Wow, thanks for going on that journey! I quite like the Fuel syntax too, it's much terser as you say. I'll try putting in the HEAD request.

Reply:
Ah so of course, hintr doesn't currently support HEAD requests (getting 404s) - I'll ask Rob to put that in.

Reply:
Ah of course. Want to ask him to add the content disposition header too?